LBType field for AWS in infra cluster object.

miheer opened the pull request:
In the ingress config cluster object we will need to add api for LBType for AWS which will store and persist the information of the desired LB type mentioned by the installer config API so that the ingress operator will refer the LB type from this object whenever someone accidentally deletes the ingress controller. Currently, the default LB type i.e CLB/ELB gets set if you delete the ingress controller object even when the desired the LB type for default ingress controller was set to NLB as ingress operator does not have a way to check the LB type as we don't store it.

Enhancement Proposal - https://github.com/openshift/enhancements/pull/1148

gcs278 commented:
/assign @rfredette

miheer commented:
@rfredette PTAL. I need to get this merged ASAP as the epic has high priority. Thanks in advance!

rfredette commented:
/lgtm

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -354,6 +354,11 @@ type AWSPlatformSpec struct {
 	ServiceEndpoints []AWSServiceEndpoint `json:"serviceEndpoints,omitempty"`
+
+	// LBType allows user to set a load balancer type.
+	// Allowed values are nlb or elb
```
Allowed values must be PascalCase to follow conventions, NLB and ELB

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -354,6 +354,11 @@ type AWSPlatformSpec struct {
+	// LBType allows user to set a load balancer type.
+	// Allowed values are nlb or elb
+	// +optional
+	LBType string `json:"lbType,omitempty"`
```
You should create an enum rather than a plain string, ie a typed string, which has some exported constants with the possible values,

Given you're adding some load balancer configuration, is there any change the user may want to be able to configure any of the load balancer options, ELB or NLB at some point in the future?

I think this probably wants to end up as a discriminated union with configuration for both types of load balancer

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -374,6 +379,11 @@ type AWSPlatformStatus struct {
 	// +optional
 	ResourceTags []AWSResourceTag `json:"resourceTags,omitempty"`
+
+	// LBType allows user to set a load balancer type.
+	// Allowed values are nlb or elb
+	// +optional
+	LBType string `json:"lbType,omitempty"`
```
Why do you need this in the status as well as the spec?

miheer commented:
@JoelSpeed PTAL. I have made the changes you asked for.

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -354,8 +354,21 @@ type AWSPlatformSpec struct {
+
+	// LBType allows user to set a load balancer type.
+	// Allowed values are NLB or Classic
+	// +kubebuilder:validation:Enum:=NLB;Classic
```
For backwards compatibility you need to allow an empty value too
```suggestion
	// +kubebuilder:validation:Enum:=NLB;Classic;""
```

JoelSpeed commented on config/v1/types_infrastructure.go:
```diff
@@ -354,8 +354,21 @@ type AWSPlatformSpec struct {
 	// +optional
 	ServiceEndpoints []AWSServiceEndpoint `json:"serviceEndpoints,omitempty"`
+
+	// LBType allows user to set a load balancer type.
```
You need to explain what happens when this field isn't set. You also need to explain what would happen if the field is changed (if that is allowed) while the cluster is running.

JoelSpeed replied:
Reading the enhancement, this seems it's the default but can be overriden right? That will need to be explained in this godoc

miheer replied:
Hi @JoelSpeed ! I have added the description for what happens when this value is set and not set.

miheer replied:
@JoelSpeed I have added the description as to what will happen when the value is set and not set.

	// When this field is set the default ingresscontroller will get created using the specified LBType.
	// If this field is not set then the default ingress controller of LBType Classic will be created.

miheer commented on config/v1/types_infrastructure.go:
```diff
@@ -354,8 +354,26 @@ type AWSPlatformSpec struct {
+
+	// LBType allows user to set a load balancer type.
+	// When this field is set the default ingresscontroller will get created using the specified LBType.
+	// If this field is not set then the default ingress controller of LBType Classic will be created.
```
The ingress operator will check this value at certain interval.

miheer commented:
@JoelSpeed based on our discussions it looks like the right place is ingress cluster object and not the infrastructure cluster object.  The default ingresscontroller can have either an internal or external LB.  However, The API uses a separate load balancer.
As you mentioned ->
"Should this be on the infrastructure object? Infrastructure to me implies things around the control plane, so i'd expect the internal and external control plane load balancers to be here, but, if this is a workload load balancer, maybe it's better suited somewhere else?"
over here https://github.com/openshift/enhancements/pull/1148#discussion_r932289883

I think ingress cluster object seems to be the right place.

I have made the necessary changes. PTAL.

Copying @sadasu from the installler team as she had voted for this field to be added in the infrastructure object.

miheer commented:
/retest

JoelSpeed commented:
Were there arguments already made for why this would be on either the ingress or infrastructure objects? Keen to see what the other perspectives were

If this is going onto the ingress object, we will want to add a PlatformSpec style discriminated union to house the data

```
spec:
  platformSpec:
    platform: AWS
    aws:
      loadBalancer:
        defaultType: NLB
```

miheer commented:
> Were there arguments already made for why this would be on either the ingress or infrastructure objects? Keen to see what the other perspectives were
> 
> If this is going onto the ingress object, we will want to add a PlatformSpec style discriminated union to house the data
> 
> ```
> spec:
>   platformSpec:
>     platform: AWS
>     aws:
>       loadBalancer:
>         defaultType: NLB
> ```

No there were not arguments made as such. But as you also mentioned that "Infrastructure to me implies things around the control plane, so i'd expect the internal and external control plane load balancers to be here, but, if this is a workload load balancer," the right place will be the ingress object as this is for workload balancer.   The default ingress controller can have either an internal or external LB.  The API uses a separate load balancer.

I will add the changes you suggested.

miheer commented:
@JoelSpeed  PTAL

miheer commented:
> > Were there arguments already made for why this would be on either the ingress or infrastructure objects? Keen to see what the other perspectives were
> > If this is going onto the ingress object, we will want to add a PlatformSpec style discriminated union to house the data
> > ```
> > spec:
> >   platformSpec:
> >     platform: AWS
> >     aws:
> >       loadBalancer:
> >         defaultType: NLB
> > ```
> 
> @JoelSpeed No there were not any arguments made as such. There was one comment from @sadasu over here [openshift/enhancements#1148 (comment)](https://github.com/openshift/enhancements/pull/1148#discussion_r919410024) . But as you also mentioned that "Infrastructure to me implies things around the control plane, so i'd expect the internal and external control plane load balancers to be here, but, if this is a workload load balancer," the right place will be the ingress object as this is for workload balancer. The default ingress controller can have either an internal or external LB. The API uses a separate load balancer.
> 
> I will quickly add the changes you suggested.

@JoelSpeed PTAL. We have a google where we had asked where to place this field. But we were asked to ask the installer team.  https://drive.google.com/file/d/1NcbZHpRWyqIvMN4Bxv6MLqVxUg_IN8ba/view?t=22m15s
However as mentioned earlier @sadasu from installer team  did vote for infrastructure object https://github.com/openshift/enhancements/pull/1148#discussion_r919410024  but as we discussed this is a workload balancer we will need to place in another object than infrastructure. So, placing this in the ingress cluster object will be the appropriate option.

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,52 @@ type IngressSpec struct {
 	// +optional
 	RequiredHSTSPolicies []RequiredHSTSPolicy `json:"requiredHSTSPolicies,omitempty"`
+
+	// platform holds desired information specific to the underlying
+	// infrastructure provider for the ingress.
+	Platform *IngressPlatformSpec `json:"platform,omitempty"`
```
Just want to reword this slightly
```suggestion
	// platform holds configuration specific to the underlying
	// infrastructure provider for the ingress load balancers.
	// When omitted, this means the user has no opinion and the platform is left
	// to choose reasonable defaults. These defaults are subject to change over time.
	// +optional
	Platform *IngressPlatformSpec `json:"platform,omitempty"`
```

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,52 @@ type IngressSpec struct {
+	Platform *IngressPlatformSpec `json:"platform,omitempty"`
+}
+
+type IngressPlatformSpec struct {
```
This is a union so needs a union tag, please also add some description of the struct
```suggestion
// +union
type IngressPlatformSpec struct {
```

JoelSpeed replied:
Please make sure to include a description of the expect use of the union in the godoc

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,52 @@ type IngressSpec struct {
+}
+
+type AWSIngressSpec struct {
+	// LoadBalancer contains the AWS load balancer details
+	LoadBalancer *LoadBalancer `json:"loadbalancer,omitempty"`
```
Is this optional? If so it needs an `+optional` tag

Also, please don't use a pointer, it harms discoverability for the end user

miheer replied:
But, how can we check for nil values in the operator code if we don't use a pointer ?
Eg- Check if Spec.Platform.LoadBalancer != nil if it is not set as it is an optional field

JoelSpeed replied:
Why do you need to check if it is nil? Do you need to distinguish between a nil value and the zero value of the struct?

miheer replied:
I just want to check if LoadBalancer is set or not. From my understanding Go will throw nil error if this value is not set as it is optional.

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,52 @@ type IngressSpec struct {
+	// Allowed values are NLB or Classic
+	// +kubebuilder:validation:Enum:=NLB;Classic;""
+	// +optional
+	DefaultLBType AWSLBType `json:"defaultLBType,omitempty"`
```
Lets make this `DefaultType`, we already know its to do with load balancers because it's in the load balancers section

The comment on this field also needs to be updated

miheer commented:
@JoelSpeed I have made some major changes as we would need the api to be built in such a way that it could handle LB option like Scope which is not platform specific and also platform and  LB Type specific settings like connection idle time out and load balancer source range.

To set LBType the api will look like this
`Spec.LoadBalancer.Platform.AWS.defaultType`

To set the scope of the Load balancer(in the future). This is not platform or LB Type specific.
`Spec.Loadbalancer.Scope`

To set the LoadBalancerSourceRange which is LBType specific(in the future)
`Spec.LoadBalancer.Platform.AWS.AWSClassicLoadBalancerParamaters.LoadBalancerSourceRange`

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,94 @@ type IngressSpec struct {
+	// +optional
+	Type AWSLBType `json:"type,omitempty"`
+
+	// classicLoadBalancerParameters holds configuration parameters for an AWS
+	// classic load balancer. Present only if type is Classic.
+	//
+	// +optional
+	ClassicLoadBalancerParameters *AWSClassicLoadBalancerParameters `json:"classicLoadBalancer,omitempty"`
+
+	// networkLoadBalancerParameters holds configuration parameters for an AWS
+	// network load balancer. Present only if type is NLB.
+	//
+	// +optional
+	NetworkLoadBalancerParameters *AWSNetworkLoadBalancerParameters `json:"networkLoadBalancer,omitempty"`
```
This looks really good, there aren't any parameters in the structs right now, I wonder if we need to include them or if we can add them later.

Probably a question for @deads2k as he knows more about the future compatibility stuff

I believe as this would be optional and all the fields in it going forward must also be optional that it's ok to omit it for now. In the future, users would set the discriminant to say `NLB` and the validation shouldn't fail if `nlb` isn't set

deads2k replied:
> I believe as this would be optional and all the fields in it going forward must also be optional that it's ok to omit it for now. In the future, users would set the discriminant to say NLB and the validation shouldn't fail if nlb isn't set

Correct, we can wait to add the struct pointers until there is configuration that is required.  The reasoning about missing and nil handling is correct.

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,94 @@ type IngressSpec struct {
+	// classic load balancer. Present only if type is Classic.
+	//
+	// +optional
+	ClassicLoadBalancerParameters *AWSClassicLoadBalancerParameters `json:"classicLoadBalancer,omitempty"`
```
The json tag must match (case insensitively) the discriminant value, so it must be `classic`, and the NLB one must be `nlb`

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,94 @@ type IngressSpec struct {
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=NLB;Classic;""
+	// +optional
```
Discriminants should be required, but the field containing the discriminant should be optional. We will need to check that it can be omitted without causing openapi validation errors

miheer replied:
@JoelSpeed Thanks for the review! Sorry I did not get the point regarding "but the field containing the discriminant should be optional"  over here ?

Do you mean I shall change the optional tag to required ?

JoelSpeed replied:
When you include `AWSIngressSpec` somewhere, that should be optional, but the discriminant (the `Type` field here) should be required

miheer commented:
@JoelSpeed I have made the changes as suggested. PTAL.

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,99 @@ type IngressSpec struct {
+
+	// AWS contains settings specific to the Amazon Web Services infrastructure provider.
+	// +optional
+	AWS AWSIngressSpec `json:"aws,omitempty"`
```
For a discriminated union, the members must be pointers, this is an exception to the normal rule

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,99 @@ type IngressSpec struct {
+	// +optional
+	Platform IngressPlatformSpec `json:"platform,omitempty"`
+
+	//TODO Add Scope
```
What does this TODO mean? Can you expand on the TODO or remove it before we merge

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,99 @@ type IngressSpec struct {
+	// +required
+	Type AWSLBType `json:"type,omitempty"`
+
+	// classicLoadBalancerParameters holds configuration parameters for an AWS
+	// classic load balancer. Present only if type is Classic.
+	//
+	// +optional
+	ClassicLoadBalancerParameters *AWSClassicLoadBalancerParameters `json:"classic,omitempty"`
+
+	// networkLoadBalancerParameters holds configuration parameters for an AWS
+	// network load balancer. Present only if type is NLB.
+	//
+	// +optional
+	NetworkLoadBalancerParameters *AWSNetworkLoadBalancerParameters `json:"nlb,omitempty"`
```
As these are empty right now, lets remove them. We can add them later when we actually have options to populate them with

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,99 @@ type IngressSpec struct {
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=NLB;Classic;""
+	// +required
```
```suggestion
	// +kubebuilder:validation:Required
```

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,99 @@ type IngressSpec struct {
+	//
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=NLB;Classic;""
```
You don't need the empty option if this is required
```suggestion
	// +kubebuilder:validation:Enum:=NLB;Classic
```

miheer commented:
I see the verify test failing for other PR's as well. After running make verify on my local machine.

Errors ->
Generating OpenAPI definitions for apiserver:v1 apps:v1 authorization:v1 build:v1 config:v1 helm:v1beta1 console:v1 console:v1alpha1 image:v1,docker10,dockerpre012 imageregistry:v1 kubecontrolplane:v1 legacyconfig:v1 cloudnetwork:v1 network:v1 networkoperator:v1 oauth:v1 openshiftcontrolplane:v1 operator:v1 operator:v1alpha1 operatorcontrolplane:v1alpha1 operatoringress:v1 osin:v1 project:v1 quota:v1 route:v1 samples:v1 security:v1 securityinternal:v1 servicecertsigner:v1alpha1 sharedresource:v1alpha1 template:v1 user:v1 machine:v1beta1 machine:v1 monitoring:v1alpha1 at github.com/openshift/api/verify_openapi
E0823 14:15:33.440417 1494307 openapi.go:487] [github.com/openshift/api/machine/v1.NutanixResourceIdentifier]: union members must be optional: github.com/openshift/api/machine/v1.NutanixResourceIdentifier.Type
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,APIRequestCountStatus,Conditions
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,APIRequestCountStatus,Last24h
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,PerNodeAPIRequestLog,ByUser
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,PerResourceAPIRequestLog,ByNode
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,PerUserAPIRequestCount,ByVerb
...
...
...
API rule violation: omitempty_match_case,github.com/openshift/api/operator/v1,StaticPodOperatorStatus,LatestAvailableRevision
API rule violation: omitempty_match_case,github.com/openshift/api/operator/v1,StaticPodOperatorStatus,LatestAvailableRevisionReason
hack/../hack/update-openapi.sh: line 43: ../../../github.com/openshift/api/verify_openapi/openapi.json: No such file or directory
make: *** [Makefile:55: verify-scripts] Error 1


Also I had to run ` go get k8s.io/kube-openapi/cmd/openapi-gen@v0.0.0-20220328201542-3ee0da9b0b42` and `go install k8s.io/kube-openapi/cmd/openapi-gen@v0.0.0-20220328201542-3ee0da9b0b42` or else it complains `openapi-gen: No such file or directory`

@JoelSpeed  @deads2k can you please let me know how I can fix the `make verify` test ?

JoelSpeed commented:
API looks good now, for the verify, try running the openapi-gen and commiting the result, `./hack/update-openapi.sh`.

You may want to try running it by picking from #1257 as this should be fairly environment detangled

miheer commented:
@JoelSpeed this is the output ->

> API looks good now, for the verify, try running the openapi-gen and commiting the result, `./hack/update-openapi.sh`.
> 
> You may want to try running it by picking from #1257 as this should be fairly environment detangled

2022/08/24 08:53:40 Arguments validation error: output package cannot be empty
[miheer@localhost api]$

[miheer@localhost api]$ ./hack/update-openapi.sh
Generating OpenAPI definitions for apiserver:v1 apps:v1 authorization:v1 build:v1 config:v1 helm:v1beta1 console:v1 console:v1alpha1 image:v1,docker10,dockerpre012 imageregistry:v1 kubecontrolplane:v1 legacyconfig:v1 cloudnetwork:v1 network:v1 networkoperator:v1 oauth:v1 openshiftcontrolplane:v1 operator:v1 operator:v1alpha1 operatorcontrolplane:v1alpha1 operatoringress:v1 osin:v1 project:v1 quota:v1 route:v1 samples:v1 security:v1 securityinternal:v1 servicecertsigner:v1alpha1 sharedresource:v1alpha1 template:v1 user:v1 machine:v1beta1 machine:v1 monitoring:v1alpha1 at github.com/openshift/api/openapi
E0824 08:49:09.173510   11128 openapi.go:487] [github.com/openshift/api/machine/v1.NutanixResourceIdentifier]: union members must be optional: github.com/openshift/api/machine/v1.NutanixResourceIdentifier.Type
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,APIRequestCountStatus,Conditions
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,APIRequestCountStatus,Last24h
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,PerNodeAPIRequestLog,ByUser
API rule violation: list_type_missing,github.com/openshift/api/apiserver/v1,PerResourceAPIRequestLog,ByNode
...
...
...
API rule violation: names_match,k8s.io/apimachinery/pkg/util/intstr,IntOrString,Type
API rule violation: omitempty_match_case,github.com/openshift/api/operator/v1,StaticPodOperatorStatus,LatestAvailableRevision
API rule violation: omitempty_match_case,github.com/openshift/api/operator/v1,StaticPodOperatorStatus,LatestAvailableRevisionReason

miheer commented:
ci/prow/verify-client-go — Job failed. @JoelSpeed @deads2k can you please check this  issue ?

./vendor/github.com/openshift/api/openapi/openapi.json
F0824 12:38:29.971600   16614 main.go:53] Error: Failed executing generator: some packages had errors:
com.github.openshift.api.config.v1.AWSIngressSpec: discriminator set to type, but no fields in union

deads2k commented:
/retest

deads2k commented:
> ci/prow/verify-client-go — Job failed. @JoelSpeed @deads2k can you please check this issue ?

we fixed the generator, you're green now.

deads2k commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,75 @@ type IngressSpec struct {
+// This only includes fields that can be modified in the cluster.
+// +union
+type AWSIngressSpec struct {
+	// DefaultType allows user to set a load balancer type.
```
DefaultType or type?

miheer replied:
sorry Type.

deads2k commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,75 @@ type IngressSpec struct {
+	//
+	//     https://docs.aws.amazon.com/AmazonECS/latest/developerguide/load-balancer-types.html#nlb
+	// +unionDiscriminator
+	// +kubebuilder:validation:Enum:=NLB;Classic
```
Will this allow ""?

miheer replied:
No we wont't allow "" as if someone is mentioning 
```	// AWS contains settings specific to the Amazon Web Services infrastructure provider.
	// +optional
	AWS *AWSIngressSpec `json:"aws,omitempty"`
```
which is optional then they shall mention the load balancer type of AWS.

JoelSpeed replied:
Required fields in optional structs that are pointers should be able to be omitted completely, so this should work. But please test the serialization and validation manually before we merge this

deads2k commented on config/v1/types_ingress.go:
```diff
@@ -82,8 +82,75 @@ type IngressSpec struct {
+	// and must handle unrecognized platforms as None if they do not support that platform.
+	//
+	// +unionDiscriminator
+	Type PlatformType `json:"type"`
```
This is repeated from the Platform.config.openshift.io so that we can have the discriminated union here?

On upgrade, this will be "", right?  Is "" allowed by validation?

miheer replied:
If the supported provider is not present in the list of // +kubebuilder:validation:Enum="";AWS;Azure;BareMetal;GCP;Libvirt;OpenStack;None;VSphere;oVirt;IBMCloud;KubeVirt;EquinixMetal;PowerVS;AlibabaCloud;Nutanix
then we shall allow ""

miheer replied:
@Miciah ^^

Miciah replied:
Right, `PlatformType` allows the empty string: https://github.com/openshift/api/blob/0f86a223d4bce1c63d9adce1f229afb08305a23b/config/v1/types_infrastructure.go#L126-L128

And we are indeed using `PlatformType` so that we can define `spec.loadBalancer.platform` as a discriminated union, making room to expand the API with platform-specific load-balancer settings for other platforms beyond the one option for the one specific platform that Miheer is adding here.

deads2k commented:
question about handling "" so that upgrades work, looks ok otherwise.  You may need to install this on a cluster to confirm.

miheer commented:
@JoelSpeed @deads2k  PTAL

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -85,8 +85,75 @@ type IngressSpec struct {
+// This only includes fields that can be modified in the cluster.
+// +union
+type AWSIngressSpec struct {
+	// Type allows user to set a load balancer type.
```
Nit, as this is user facing, the comment should match the json tag not go type
```suggestion
	// type allows user to set a load balancer type.
```

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -85,8 +85,75 @@ type IngressSpec struct {
+	// +unionDiscriminator
+	Type PlatformType `json:"type"`
+
+	// AWS contains settings specific to the Amazon Web Services infrastructure provider.
```
Nit, as this is user facing, the comment should match the json tag not go type
```suggestion
	// aws contains settings specific to the Amazon Web Services infrastructure provider.
```

JoelSpeed commented on config/v1/types_ingress.go:
```diff
@@ -85,8 +85,75 @@ type IngressSpec struct {
 	// +optional
 	RequiredHSTSPolicies []RequiredHSTSPolicy `json:"requiredHSTSPolicies,omitempty"`
+
+	// LoadBalancer contains the AWS load balancer details
```
Nit, as this is user facing, the comment should match the json tag not go type
```suggestion
	// loadBalancer contains the AWS load balancer details
```

While this struct currently only contains aws config, maybe we should expand the comment so we don't need to update it when we add new platforms in the future?

deads2k commented:
/approve
/hold

holding for updates to Joel's comments.  The API lgtm otherwise.

miheer commented:
@JoelSpeed PTAL

miheer commented:
/retest

miheer commented:
@JoelSpeed any idea why the verify is failing ?

+ /go/src/github.com/openshift/api/hack/update-deepcopy.sh
Generating deepcopy funcs
F0826 02:04:41.618098    8762 main.go:80] Error: Failed executing generator: some packages had errors:
errors in package "github.com/openshift/api/config/v1":
output for "v1/zz_generated.deepcopy.go" differs; first existing/expected diff: 
  "PlatformSpec) DeepCopyInto(out *AWSPlatformSpec) {\n\t*out = *in\n\tif in.ServiceEndpoints != nil {\n\t\tin"
  "IngressSpec) DeepCopyInto(out *AWSIngressSpec) {\n\t*out = *in\n\treturn\n}\n\n// DeepCopy is an autogenera"


on my local machine make verify works with no errors. I have generated and pushed all the files.

JoelSpeed commented:
@miheer There are other changes in master that affect this deepcopy file since your branch was created, if you rebase on latest master and then run the deepcopy generator you'll see there is a diff

JoelSpeed commented:
Thanks @miheer, this looks good. Could you please manually test the new API and verify that the validations are working as expected (eg the optional fields can be omitted, the enums work correctly), then I'm happy to merge this

miheer commented:
@JoelSpeed I was able to test this and it seems to be working fine

# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: config.openshift.io/v1
kind: Ingress
metadata:
  creationTimestamp: "2022-08-30T01:03:17Z"
  generation: 11
  name: cluster
  resourceVersion: "60888"
  uid: de89cc95-449e-4ae5-a837-f4171fe61e36
spec:
  domain: apps.testnlbapi.devcluster.openshift.com
  loadbalancer:
    platform:
      aws:
        type: NLB
      type: AWS
status:
  componentRoutes:

works fine ->
removed loadbalancer It is not shown in the CR after saving the CR
removed platform It is not shown in the CR after saving the CR
removed aws It is not shown in the CR after saving the CR
removed NLB ....told to add a value
removed type: AWS ....It is not shown in the CR after saving the CR (Is this the right behavior ?)

JoelSpeed commented:
Great thanks! 

/lgtm

JoelSpeed commented:
/hold cancel

miheer commented:
@ahardin-rh can you please docs-approved this PR ?
@hongli can you please qe-approved this PR ?

lihongan commented:
/label qe-approved

CFields651 commented:
/label px-approved

ahardin-rh commented:
/label docs-approved
